File: sop-sdk/sdk-java/src/main/java/com/gitee/sop/sdk/client/OpenClient.java
package com.gitee.sop.sdk.client;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.gitee.sop.sdk.common.DataNameBuilder;
import com.gitee.sop.sdk.common.OpenConfig;
import com.gitee.sop.sdk.common.RequestForm;
import com.gitee.sop.sdk.exception.SdkException;
import com.gitee.sop.sdk.request.BaseRequest;
import com.gitee.sop.sdk.response.BaseResponse;
import com.gitee.sop.sdk.sign.SopSignException;
import com.gitee.sop.sdk.sign.SopSignature;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import java.util.Collections;
import java.util.Map;

/**
 * 请求客户端
 *
 * @author tanghc
 */
public class OpenClient {
    private static final Log log = LogFactory.getLog(OpenClient.class);

    private static final OpenConfig DEFAULT_CONFIG = new OpenConfig();

    private String url;
    private String appId;
    private String privateKey;

    private OpenConfig openConfig;
    private OpenRequest openRequest;
    private DataNameBuilder dataNameBuilder;

    public OpenClient(String url, String appId, String privateKey) {
        this(url, appId, privateKey, DEFAULT_CONFIG);
    }

    public OpenClient(String url, String appId, String privateKey, OpenConfig openConfig) {
        if (openConfig == null) {
            throw new IllegalArgumentException("openConfig不能为null");
        }
        this.url = url;
        this.appId = appId;
        this.privateKey = privateKey;
        this.openConfig = openConfig;

        this.openRequest = new OpenRequest(openConfig);
        this.dataNameBuilder = openConfig.getDataNameBuilder();
    }

    /**
     * 请求接口
     *
     * @param request 请求对象
     * @param <T>     返回对应的Response
     * @return 返回Response
     */
    public <T extends BaseResponse> T execute(BaseRequest<T> request) {
        return this.execute(request, null);
    }

    /**
     * 请求接口
     *
     * @param request     请求对象
     * @param accessToken jwt
     * @param <T>         返回对应的Response
     * @return 返回Response
     */
    public <T extends BaseResponse> T execute(BaseRequest<T> request, String accessToken) {
        RequestForm requestForm = request.createRequestForm(this.openConfig);
        // 表单数据
        Map<String, String> form = requestForm.getForm();
        if (accessToken != null) {
            form.put(this.openConfig.getAccessTokenName(), accessToken);
        }
        form.put(this.openConfig.getAppKeyName(), this.appId);

        String content = SopSignature.getSignContent(form);
        String sign = null;
        try {
            sign = SopSignature.rsa256Sign(content, privateKey, "utf-8");
        } catch (SopSignException e) {
            throw new SdkException("构建签名错误", e);
        }

        form.put(this.openConfig.getSignName(), sign);

        String resp = doExecute(this.url, requestForm, Collections.emptyMap());
        if (log.isDebugEnabled()) {
            log.debug("----------- 请求信息 -----------"
                    + "\n请求参数：" + SopSignature.getSignContent(form)
                    + "\n待签名内容：" + content
                    + "\n签名(sign)：" + sign
                    + "\n----------- 返回结果 -----------"
                    + "\n" + resp
            );
        }
        return this.parseResponse(resp, request);
    }

    protected String doExecute(String url, RequestForm requestForm, Map<String, String> header) {
        return openRequest.request(url, requestForm, header);
    }

    protected <T extends BaseResponse> T parseResponse(String resp, BaseRequest<T> request) {
        String method = request.getMethod();
        String dataName = dataNameBuilder.build(method);
        JSONObject jsonObject = JSON.parseObject(resp);
        JSONObject data = jsonObject.getJSONObject(dataName);
        T t = data.toJavaObject(request.getResponseClass());
        t.setBody(data.toJSONString());
        return t;
    }


}

File: sop-sdk/sdk-csharp/SDKCSharp/Client/OpenClient.cs
﻿using System;
using System.Web;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using System.Threading.Tasks;

using SDKCSharp.Common;
using SDKCSharp.Request;
using SDKCSharp.Response;
using SDKCSharp.Utility;
using System.IO;
using Newtonsoft.Json.Linq;

namespace SDKCSharp.Client
{
    /// <summary>
    /// 客户端
    /// </summary>
    public class OpenClient
    {

        private static OpenConfig DEFAULT_CONFIG = new OpenConfig();

        private Dictionary<string, string> header = new Dictionary<string, string>();


        private string url;
        private string appId;
        private string privateKey;
        private string publicKeyPlatform;

        private OpenConfig openConfig;
        private OpenRequest openRequest;
        private DataNameBuilder dataNameBuilder;


        public OpenClient(string url, string appId, string privateKey) 
            : this(url, appId, privateKey,false, DEFAULT_CONFIG)
        {
            
        }

        public OpenClient(string url, string appId, string privateKey, string publicKeyPlatform)
            : this(url, appId, privateKey)
        {
            this.publicKeyPlatform = publicKeyPlatform;
        }

        public OpenClient(string url, string appId, string privateKey, bool priKeyFromFile) 
            : this(url, appId, privateKey, priKeyFromFile, DEFAULT_CONFIG)
        {

        }

        public OpenClient(string url, string appId, string privateKey, bool priKeyFromFile, string publicKeyPlatform)
            : this(url, appId, privateKey, priKeyFromFile)
        {
            this.publicKeyPlatform = publicKeyPlatform;
        }

        public OpenClient(string url, string appId, string privateKey,bool priKeyFromFile, OpenConfig openConfig)
        {
            this.url = url;
            this.appId = appId;
            this.privateKey = privateKey;
            this.openConfig = openConfig;
            this.openRequest = new OpenRequest(openConfig);
            // 如果是从文件中加载私钥
            if (priKeyFromFile)
            {
                this.privateKey = LoadCertificateFile(privateKey);
            }
            this.dataNameBuilder = openConfig.DataNameBuilder;
        }

        public OpenClient(string url, string appId, string privateKey, bool priKeyFromFile, string publicKeyPlatform, OpenConfig openConfig)
            : this(url, appId, privateKey, priKeyFromFile, openConfig)
        {
            this.publicKeyPlatform = publicKeyPlatform;
        }

        /// <summary>
        /// 加载秘钥文件
        /// </summary>
        /// <returns>返回私钥内容.</returns>
        /// <param name="filename">文件路径.</param>
        private static string LoadCertificateFile(string filename)
        {
            if(!File.Exists(filename))
            {
                throw new SopException("文件不存在," + filename);
            }
            using (FileStream fs = File.OpenRead(filename))
            {
                byte[] data = new byte[fs.Length];
                fs.Read(data, 0, data.Length);
                return Encoding.UTF8.GetString(data);
            }
        }

        /// <summary>
        /// 发送请求
        /// </summary>
        /// <typeparam name="T">返回的Response类</typeparam>
        /// <param name="request">请求对象</param>
        /// <returns>返回Response类</returns>
        public virtual T Execute<T>(BaseRequest<T> request) where T : BaseResponse
        {
            return this.Execute<T>(request, null);
        }

        /// <summary>
        /// 发送请求
        /// </summary>
        /// <typeparam name="T">返回的Response类</typeparam>
        /// <param name="request">请求对象</param>
        /// <param name="accessToken">accessToken</param>
        /// <returns>返回Response类</returns>
        public virtual T Execute<T>(BaseRequest<T> request, string accessToken) where T : BaseResponse
        {
            RequestForm requestForm = request.CreateRequestForm(this.openConfig);
            Dictionary<string, string> form = requestForm.Form;
            if (!string.IsNullOrEmpty(accessToken))
            {
                form[this.openConfig.AccessTokenName] = accessToken;
            }
            form[this.openConfig.AppKeyName] = this.appId;
            string sign = SignUtil.CreateSign(form, privateKey, openConfig.Charset, openConfig.SignType);
            form[this.openConfig.SignName] = sign;

            string resp = this.DoExecute(url, requestForm, header);

            return this.ParseResponse<T>(resp, request);
        }

        /// <summary>
        /// 执行请求
        /// </summary>
        /// <param name="url">请求url</param>
        /// <param name="requestForm">请求内容</param>
        /// <param name="header">请求header</param>
        /// <returns>返回服务器响应内容</returns>
        protected virtual string DoExecute(string url, RequestForm requestForm, Dictionary<string, string> header)
        {
            return openRequest.Request(this.url, requestForm, header);
        }

        /// <summary>
        /// 解析返回结果
        /// </summary>
        /// <typeparam name="T">返回的Response</typeparam>
        /// <param name="resp">服务器响应内容</param>
        /// <param name="request">请求Request</param>
        /// <returns>返回Response</returns>
        protected virtual T ParseResponse<T>(string resp, BaseRequest<T> request) where T: BaseResponse
        {
            string method = request.Method;
            string rootNodeName = this.dataNameBuilder.Build(method);
            string errorRootNode = openConfig.ErrorResponseName;
            Dictionary<string, object> responseData = JsonUtil.ParseToDictionary(resp);
            bool errorResponse = responseData.ContainsKey(errorRootNode);
            if (errorResponse)
            {
                rootNodeName = errorRootNode;
            }
            object data = responseData[rootNodeName];
            responseData.TryGetValue(openConfig.SignName, out object sign);
            if (sign != null && !string.IsNullOrEmpty(publicKeyPlatform))
            {
                string signContent = BuildBizJson(rootNodeName, resp);
                if (!CheckResponseSign(signContent, sign.ToString(), publicKeyPlatform))
                {
                    ErrorResponse checkSignErrorResponse = SopSdkErrors.CHECK_RESPONSE_SIGN_ERROR;
                    data = JsonUtil.ToJSONString(checkSignErrorResponse);
                }
            }
            string jsonData = data == null ? "{}" : data.ToString();
            T t = JsonUtil.ParseObject<T>(jsonData);
            t.Body = jsonData;
            return t;
        }

        /// <summary>
        /// 验证服务端返回的sign
        /// </summary>
        /// <returns><c>true</c>, if response sign was checked, <c>false</c> otherwise.</returns>
        /// <param name="signContent">Response data.</param>
        /// <param name="sign">sign data.</param>
        /// <param name="publicKeyPlatform">Public key platform.</param>
        protected virtual bool CheckResponseSign(string signContent, string sign, string publicKeyPlatform)
        {
            try
            {
                Encoding charset = openConfig.Charset;
                SignType signType = openConfig.SignType;
                return SignUtil.RsaCheck(signContent, sign, publicKeyPlatform, charset, signType);
            }
            catch (Exception)
            {
                return false;
            }
        }

        protected virtual string BuildBizJson(string rootNodeName, string body)
        {
            int indexOfRootNode = body.IndexOf(rootNodeName);
            if (indexOfRootNode < 0)
            {
                rootNodeName = openConfig.ErrorResponseName;
                indexOfRootNode = body.IndexOf(rootNodeName);
            }
            string result = null;
            if (indexOfRootNode > 0)
            {
                result = BuildJsonNodeData(body, rootNodeName, indexOfRootNode);
            }
            return result;
        }

        protected virtual string BuildJsonNodeData(string body, string rootNode, int indexOfRootNode)
        {
            int signDataStartIndex = indexOfRootNode + rootNode.Length + 2;
            int indexOfSign = body.IndexOf("\"" + openConfig.SignName  + "\"");
            if (indexOfSign < 0)
            {
                return null;
            }

            int signDataEndIndex = indexOfSign - 1;
            int length = signDataEndIndex - signDataStartIndex;

            return body.Substring(signDataStartIndex, length);
        }
    }
}
